clang/lib/AST/PreorderAST.cpp

```diff
@@ -56,11 +62,58 @@ void PreorderAST::Create(Expr *E, Node *Parent) {
   E = Lex.IgnoreValuePreservingOperations(Ctx, E->IgnoreParens());
 
   if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
-    auto *N = new BinaryNode(BO->getOpcode(), Parent);
+    BinaryOperator::Opcode BinOp = BO->getOpcode();
+    Expr *LHS = BO->getLHS();
+    Expr *RHS = BO->getRHS();
+
+    // We can convert (e1 - e2) to (e1 + -e2) if -e2 does not overflow.  One
+    // instance where -e2 can overflow is if e2 is INT_MIN. Here, instead of
+    // specifically checking whether e2 is INT_MIN, we add a unary minus to e2
+    // and then check if the resultant expression -e2 overflows. If it
+    // overflows, we undo the unary minus operator.
+
+    // TODO: Currently we only handle detection of overflow for constant
+    // expressions. We need to handle the case where e2 is a variable
+    // expression which is not known until run time.
+    if (BO->getOpcode() == BO_Sub) {
+      RHS = new (Ctx) UnaryOperator(RHS, UO_Minus, RHS->getType(),
+                                    RHS->getValueKind(),
+                                    RHS->getObjectKind(),
+                                    SourceLocation(),
+                                    /*CanOverflow*/ true);
+      bool Overflow = false;
+      if (RHS->isIntegerConstantExpr(Ctx)) {
+        SmallVector<PartialDiagnosticAt, 8> Diag;
+        llvm::APSInt IntVal =
+          RHS->EvaluateKnownConstIntCheckOverflow(Ctx, &Diag);
+
+        for (auto &PD : Diag) {
+          if (PD.second.getDiagID() == diag::note_constexpr_overflow) {
+            Overflow = true;
+            break;
+          }
+        }
+      }
+
+      if (Overflow)
+        RHS = dyn_cast<UnaryOperator>(RHS)->getSubExpr();
```

Comment:
This is more complex than needed.  How about saving the original value of RHS?

Reply:
I was trying to avoid leaking memory here. If we save the original value of RHS which is an Expr we would not be able to reclaim memory because the destructor for Expr is protected.